README.md

```diff
@@ -29,3 +29,84 @@ It is appreciated, though not required, if a Dockerfile is included.
+[]'
+```
+
+`POST` some favourites for this user. To post a favourite, include a uuid (random), description (whatever you like) and resource type (audience|chart|insight):
+```
+$  curl -f 'http://localhost:8086/favourites' -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJmYXZvdXJpdGVzIiwidXNlcmlkIjoiNjA5ZGFjOWMtYWM3OS00ZGM4LWExZjUtZjJhZjdhNTUxOWNNmIiwiaWF0IjoxNzI4MDM2ODYyLCJleHAiOjE3MzIwNTA0NjJ9.19NYScvE4f6FIHAojMcn0sv-xOgzrCE26jFsTqddoOM" -X POST -H 'Content-Type: application/json' -d '{"id": "5791fb0a-130d-4150-835a-d0433bf6eb73", "description": "my favourite chart", "resourceType": "chart"}'
+{"Description":"my favourite chart","ResourceType":"chart","Id":"5791fb0a-130d-4150-835a-d0433bf6eb73"}'
+
+$ curl -f  'http://localhost:8086/favourites' -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJmYXZvdXJpdGVzIiwidXNlcmlkIjoiNjA5ZGFjOWMtYWM3OS00ZGM4LWExZjUtZjJhZjdhNTUxOWNmIiwiaWF0IjoxNzI4MDM2ODYyLCJleHAiOjE3MzIwNTA0NjJ9.19NYScvE4f6FIHAojMcn0sv-xOgzrCE26jFsTqddoOM" -X POST -H 'Content-Type: application/json' -d '{"id": "19054c1a-5a10-43e1-aef1-14a37fddb351", "description": "my favourite insight", "resourceType": "insight"}'
+{"Description":"my favourite insight","ResourceType":"insight","Id":"19054c1a-5a10-43e1-aef1-14a37fddb351"}'
+```
+
+`GET` the user's favourites (chart and insight data are generated randomly, but IDs are the ones supplied before):
+```
+$ curl -f 'http://localhost:8086/favourites?offset=0&pagesize=10' -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJmYXZvdXJpdGVzIiwidXNlcmlkIjoiNjA5ZGFjOWMtYWM3OS00ZGM4LWExZjUtZjJhZjdhNTUxOWNmIiwiaWF0IjoxNzI4MDM2ODYyLCJleHAiOjE3MzIwNTA0NjJ9.19NYScvE4f6FIHAojMcn0sv-xOgzrCE26jFsTqddoOM"
+[
+    {
+        "Description": "my favourite chart",
+        "Chart": {
+            "Id": "5791fb0a-130d-4150-835a-d0433bf6eb73",
+            "Title": "sales chart number 5",
+            "XAxisTitle": "time",
+            "YAxisTitle": "number of sales",
+            "DataPoints": [
+                {
+                    "X": 0,
+                    "Y": 0
+                },
+                {
+                    "X": 1,
+                    "Y": 2
+                },
+                {
+                    "X": 2,
+                    "Y": 4
+                }
+            ]
+        }
+    },
+    {
+        "Description": "my favourite insight",
+        "Insight": {
+            "Id": "19054c1a-5a10-43e1-aef1-14a37fddb351",
+            "Description": "70% of people between 38 and 49 spend more than 5 hours per day online"
+        }
+    }
+]'
+
+For full API documentation, visit `http://localhost:8086/swagger/index.html` when the service is running.
+
+```
+
+## About the implementation
+I chose to implement the web server without any data persistence. 
```

Comment:
If we told you to use a Database system instead of keeping all data in memory
- Which kind of database would you choose and why?
- Can you provide a brief schema of this database (tables definition, relationships, or whatever you would do) ?
- Would you use any ORM to connect and exchange data with the database?

Reply:
### Database type
The kind of database would depend on the non-functional requirements for the project. 
What scalability would it need? Would it be read-heavy or write-heavy? etc.
Having said that, I'd probably default to whatever RDBMS is now being used at the company, which I'd expect to be Postgres, for most services here and only introduce other storage systems if there's a need, and after careful consideration.

### Schema
Let's say favourites, audiences and insights exist in the same DB while charts come from a separate service with its own DB.

#### Main service (favourites-audiences-insights):
```
Audience
- ID uuid (PK)
- Gender: char
- BirthCountry string (FK -> countries.CountryCode)
- AgeFrom int
- AgeTo int
- HoursSpentFrom int
- HoursSpentTo int
- PurchasesFrom int
- PurchasesTo int
* if there were more properties and they don't always have values, they'd probably go into a separate table instead of continuously adding columns.
```
```
Country
- CountryCode varchar (PK)
- Description varchar
```
```
Insight
- ID uuid (PK)
- Description varchar
```
```
User
- ID uuid (PK)
- FirstName varchar
- LastName varchar
- etc etc
```
```
UserFavourite
- ID uuid (PK)
- UserID uuid (FK -> User.Id)
- Description
- ResourceType varchar
- FavouriteId uuid
```
I would probably NOT apply a foreign key constraint on FavouriteId since (a) it's going to refer to different types of entities, some of which (charts) don't exist in this DB and (b) it would potentially affect write performance on this table. 

#### Charts service (its own DB - SQL now but might fit in a NoSQL / document database nicely)

```
Chart
- ID uuid (PK)
- Title string
- XAxisTitle string
- YAxisTitle 
```

```
ChartDataPoint
- ID uuid (PK)
- ChartID uuid (FK -> Chart.ID)
- X float
- Y float
```

Re. the use of an ORM, I'd probably go with whatever is being used at the company, and keep an eye on the actual queries being executed on the RDBMS.